File: src/solutions/2023/07/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

var ranking = map[byte]int{
	'A': 13,
	'K': 12,
	'Q': 11,
	'J': 10,
	'T': 9,
	'9': 8,
	'8': 7,
	'7': 6,
	'6': 5,
	'5': 4,
	'4': 3,
	'3': 2,
	'2': 1,
	'1': 0,
}

const (
	highCard int = iota
	onePair
	twoPair
	threeOfAKind
	fullHouse
	fourOfAKind
	fiveOfAKind
)

type pair struct {
	card  string
	value int
}

func sortCard(a pair, b pair) int {
	for i := 0; i < 5; i++ {
		if a.card[i] != b.card[i] {
			return ranking[a.card[i]] - ranking[b.card[i]]
		}
	}
	return 0
}

func filterzero(values [14]int) []int {
	var newAry []int

	for i := 0; i < 14; i++ {
		if values[i] != 0 {
			newAry = append(newAry, values[i])
		}
	}

	return newAry
}

func indexOf(values [14]int, value int) int {
	for i := 0; i < 14; i++ {
		if values[i] == value {
			return i
		}
	}
	return -1
}

func maxAry(values [14]int) int {
	max := values[0]
	for _, v := range values {
		if v > max {
			max = v
		}
	}
	return max
}

func getType(str string) int {
	var values [14]int
	for i := range str {
		values[ranking[str[i]]]++
	}
	if values[0] > 0 {
		var temp = values[0]
		values[0] = 0
		var idx = indexOf(values, maxAry(values))
		values[idx] += temp
	}

	newValues := filterzero(values)
	if len(newValues) == 1 {
		return fiveOfAKind
	}
	if len(newValues) == 4 {
		return onePair
	}
	if len(newValues) == 5 {
		return highCard
	}

	mn := 5
	mx := 0
	for _, v := range newValues {
		if mn > v {
			mn = v
		}
		if mx < v {
			mx = v
		}
	}
	if mn == 1 {
		if mx == 2 {
			return twoPair
		}
		if mx == 3 {
			return threeOfAKind
		}
		if mx == 4 {
			return fourOfAKind
		}
	}
	return fullHouse
}

func parseInput(input string, joker bool) [7][]pair {
	var groups [7][]pair

	for _, line := range strings.Split(input, "\n") {
		var temp []string = strings.Split(line, " ")
		if joker {
			temp[0] = strings.ReplaceAll(temp[0], "J", "1")
		}
		val, _ := strconv.ParseInt(temp[1], 10, 32)
		p := pair{temp[0], int(val)}
		t := getType(temp[0])
		groups[t] = append(groups[t], p)
	}

	return groups
}

func solve(input string, joker bool) string {
	groups := parseInput(input, joker)
	res := 0
	i := 1
	for _, group := range groups {
		sort.Slice(group, func(i, j int) bool {
			return sortCard(group[i], group[j]) < 0
		})
		for _, pair := range group {
			res += pair.value * i
			i++
		}
	}
	return strconv.Itoa(res)
}

func part1(input string, isTest bool) string {
	return solve(input, false)
}

func part2(input string, isTest bool) string {
	return solve(input, true)
}

File: src/solutions/2023/07/python/main.py
import sys
import os
import re
from typing import List
from enum import Enum

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': False, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


ranking = {
    'A': 13,
    'K': 12,
    'Q': 11,
    'J': 10,
    'T': 9,
    '9': 8,
    '8': 7,
    '7': 6,
    '6': 5,
    '5': 4,
    '4': 3,
    '3': 2,
    '2': 1,
    '1': 0,
}

card_type = {
    'HIGH_CARD': 0,
    'ONE_PAIR': 1,
    'TWO_PAIR': 2,
    'THREE_OF_A_KIND': 3,
    'FULL_HOUSE': 4,
    'FOUR_OF_A_KIND': 5,
    'FIVE_OF_A_KIND': 6,
}


def sort_card(a: [str, int]) -> int:
   return [ranking[i] for i in a[0]]


def get_type(str: str) -> int:
   values = [0 for _ in range(14)]
   for k in str:
      values[ranking[k]] += 1
   if (values[0]):
      temp = values[0]
      values[0] = 0
      idx = values.index(max(values))
      values[idx] += temp

   values = [x for x in values if x > 0]
   l = len(values)
   if l == 1: return card_type['FIVE_OF_A_KIND']
   if l == 4: return card_type['ONE_PAIR']
   if l == 5: return card_type['HIGH_CARD']

   mn = 5
   mx = 0
   for v in values:
      if mn > v: mn = v
      if mx < v: mx = v
   if mn == 1:
      if mx == 2: return card_type['TWO_PAIR']
      if mx == 3: return card_type['THREE_OF_A_KIND']
      if mx == 4: return card_type['FOUR_OF_A_KIND']
   return card_type['FULL_HOUSE']


def parse_input(input: str, joker: bool) -> [str, int]:
   temp = input.split(' ')
   if joker:
      temp[0] = temp[0].replace('J', '1')
   return [temp[0], int(temp[1])]


def solve(input: str, joker: bool) -> str:
   groups = [[] for _ in range(7)]
   for y in [parse_input(x, joker) for x in input.splitlines()]:
      groups[get_type(y[0])].append(y)

   res = 0
   i = 1
   for g in groups:
      for s in sorted(g, key=sort_card):
         res += s[1] * i
         i += 1
   return str(res)


def part_one(input: str, is_test: bool) -> str:
   return solve(input, False)


def part_two(input: str, is_test: bool) -> str:
   return solve(input, True)


if __name__ == '__main__':
   main()
